controller/pkg/packet/packet.go

```diff
@@ -57,138 +59,165 @@ func New(context uint64, bytes []byte, mark string, lengthValidate bool) (packet
+
+	if p.IPHdr.ipHeaderLen != minIPv4HdrSize {
+		return fmt.Errorf("packets with ip options not supported: hdrlen=%d", p.IPHdr.ipHeaderLen)
 	}
 
-	// Some sanity checking for UDP.
-	if p.IPProto == IPProtocolUDP {
-		if p.IPTotalLength < minUDPIPPacketLen {
-			return nil, fmt.Errorf("udp ip packet too small: hdrlen=%d", p.ipHeaderLen)
+	p.IPHdr.Buffer = bytes
+
+	if lengthValidate && p.IPHdr.ipTotalLength != uint16(len(p.IPHdr.Buffer)) {
+		if p.IPHdr.ipTotalLength < uint16(len(p.IPHdr.Buffer)) {
+			p.IPHdr.Buffer = p.IPHdr.Buffer[:p.IPHdr.ipTotalLength]
+		} else {
+			return fmt.Errorf("stated ip packet length %d differs from bytes available %d", p.IPHdr.ipTotalLength, len(p.IPHdr.Buffer))
 		}
 	}
 
-	if lengthValidate && p.IPTotalLength != uint16(len(p.Buffer)) {
-		if p.IPTotalLength < uint16(len(p.Buffer)) {
-			p.Buffer = p.Buffer[:p.IPTotalLength]
-		} else {
-			return nil, fmt.Errorf("stated ip packet length %d differs from bytes available %d", p.IPTotalLength, len(p.Buffer))
+	// Some sanity checking for TCP.
+	if p.IPHdr.ipProto == IPProtocolTCP {
+		if p.IPHdr.ipTotalLength-uint16(p.IPHdr.ipHeaderLen) < minTCPIPPacketLen {
+			return fmt.Errorf("tcp ip packet too small: hdrlen=%d", p.IPHdr.ipHeaderLen)
 		}
+
+		p.parseTCP(bytes)
 	}
 
-	if p.IPProto == IPProtocolTCP {
-		// TCP Header Processing
-		p.l4BeginPos = minIPHdrSize
-		p.TCPChecksum = binary.BigEndian.Uint16(bytes[TCPChecksumPos : TCPChecksumPos+2])
-		p.SourcePort = binary.BigEndian.Uint16(bytes[tcpSourcePortPos : tcpSourcePortPos+2])
-		p.DestinationPort = binary.BigEndian.Uint16(bytes[tcpDestPortPos : tcpDestPortPos+2])
-		p.TCPAck = binary.BigEndian.Uint32(bytes[tcpAckPos : tcpAckPos+4])
-		p.TCPSeq = binary.BigEndian.Uint32(bytes[tcpSeqPos : tcpSeqPos+4])
-		p.tcpDataOffset = (bytes[tcpDataOffsetPos] & tcpDataOffsetMask) >> 4
-		p.TCPFlags = bytes[tcpFlagsOffsetPos]
-
-		// Options and Payload that maybe added
-		p.tcpOptions = []byte{}
-		p.tcpData = []byte{}
-	} else {
-		// UDP Header Processing
-		p.UDPChecksum = binary.BigEndian.Uint16(bytes[UDPChecksumPos : UDPChecksumPos+2])
-		p.udpData = []byte{}
-		p.SourcePort = binary.BigEndian.Uint16(bytes[tcpSourcePortPos : tcpSourcePortPos+2])
-		p.DestinationPort = binary.BigEndian.Uint16(bytes[tcpDestPortPos : tcpDestPortPos+2])
+	// Some sanity checking for UDP.
+	if p.IPHdr.ipProto == IPProtocolUDP {
+		if p.IPHdr.ipTotalLength-uint16(p.IPHdr.ipHeaderLen) < minUDPIPPacketLen {
+			return fmt.Errorf("udp ip packet too small: hdrlen=%d", p.IPHdr.ipHeaderLen)
+		}
+
+		p.parseUDP(bytes)
 	}
-	p.context = context
 
-	return &p, nil
+	return nil
+}
+
+func (p *Packet) parseIPv6Packet() (packet *Packet, err error) {
+	return nil, nil
 }
 
 // IsEmptyTCPPayload returns the TCP data offset
 func (p *Packet) IsEmptyTCPPayload() bool {
-	return p.TCPDataStartBytes() == p.IPTotalLength
+	return p.TCPDataStartBytes() == uint16(len(p.IPHdr.Buffer[p.IPHdr.ipHeaderLen:]))
```

Comment:
One change that I made is all offsets are calculated from the beginning of their respective header. TCPDataStartBytes() is the offset from the beginning of the TCP packet and not from the start of the IP packet. 
I will create a new tcpTotalLength variable at the time of parsing packet which can be used here instead of finding the length of the buffer which is happening above. I will fix this.